### PR TITLE
Implement hash partitioned aggregation

### LINKED ISSUE #27: Implement hash-partitioned hash aggregate
This is (up to 4x in my earlier tests) faster than the current implementation that collects all parts to one "full" for cases with very high cardinality in the aggregate (think deduplication code). However, not hash partitioning is faster for very "simple" aggregates as less work needs to be done.

We probably need some fast way to have a rough estimate on the number of distinct values in the aggregate keys, maybe dynamically based on the first batch(es).

Also this work creates a building block for ballista to distribute data across workers, parallelizing it, avoiding collecting it to one worker, and making it scale to bigger datasets.

### PR DESCRIPTION
**Pull request by Dandandan**
# Which issue does this PR close?

Closes #27

 # Rationale for this change

A more scalable hash aggregate that works well for group by expressions that have high cardinality int he output and allows to scale better with the number of cpu cores.
The algorithm changes the steps from:

`partial hash aggregate -> merge (to 1 partition) -> full hash aggregate`

to

`partial hash aggregate -> repartition on group by expressions -> hash aggregate (on partitions)`

This is the same as what Spark is doing.

This mostly has an effect on group by with higher cardinality, but no substantial effect on lower cardinality (as partial result is already small).
For Ballista this would also be required I think @andygrove  - currently every partition is merged into one which can be problematic (and slow).

For example, TPC-H query 3 benefits quite a bit: 
Master: 
```
Query 3 avg time: 71.82 ms
```

PR:
```
Query 3 avg time: 49.20 ms
```

The db-benchmark group by queries improve by quite a bit:

Master:

```
q1 took 34 ms
q2 took 454 ms
q3 took 3284 ms
q4 took 42 ms
q5 took 2924 ms
q7 took 2843 ms
```

PR
```
q1 took 33 ms
q2 took 369 ms
q3 took 1875 ms
q4 took 46 ms
q5 took 1756 ms
q7 took 1686 ms
```

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: ballista/rust/scheduler/src/test_utils.rs
Comment (andygrove) [inline ballista/rust/scheduler/src/test_utils.rs:36]: Why is CoalesceBatches removed here?

Comment (Dandandan) [reply ballista/rust/scheduler/src/test_utils.rs:36]: Good catch - will put it back

File: datafusion/src/physical_plan/planner.rs
Comment (Dandandan) [inline datafusion/src/physical_plan/planner.rs:188]: Will create an issue for this

File: datafusion/src/execution/context.rs
Comment (alamb) [inline datafusion/src/execution/context.rs:1308]: this is a good change (to remove the redundant assertion)

File: datafusion/src/physical_plan/planner.rs
Comment (alamb) [reply datafusion/src/physical_plan/planner.rs:188]: thank you

File: datafusion/src/physical_plan/hash_aggregate.rs
Comment (alamb) [inline datafusion/src/physical_plan/hash_aggregate.rs:79]:
```suggestion
    /// Final aggregate that works on pre-partitioned data.
    ///
    /// This requires the invariant that all rows with a particular
    /// grouping key are in the same partitions, such as is the case
    /// with Hash repartitioning on the group keys. If a group key is
    /// duplicated, duplicate groups would be produced 
```

File: datafusion/src/physical_plan/hash_aggregate.rs
Comment (alamb) [inline datafusion/src/physical_plan/hash_aggregate.rs:207]: It seems like the required  required distribution for `FinalPartitioned` is really "hash partitioned" or some other partitioning that guarantees the same group key doesn't appear in multiple partitions. Leaving it as `UnspecifiedDistribution` might cause subtle bugs in the future.

File: datafusion/src/physical_plan/hash_join.rs
Comment (alamb) [inline datafusion/src/physical_plan/hash_join.rs:1031]: Did you have to add these match arms because tests failed without doing so?

**Review by alamb — APPROVED** [review]
This looks really cool @Dandandan. I suggest you add a test, if possible, that shows the plan with the repartition exec operation in order to prevent someone accidentally turning off this optimization during a refactor

File: datafusion/src/physical_plan/hash_aggregate.rs
Comment (Dandandan) [reply datafusion/src/physical_plan/hash_aggregate.rs:207]: That's interesting - we could I think add a `Distribution::HashPartitioned(Vec<Expr>)` variant, which also could be used later for optimizations.
Currently only `SinglePartition` partition is used for the `MergeExec` rule.

File: datafusion/src/physical_plan/hash_join.rs
Comment (Dandandan) [reply datafusion/src/physical_plan/hash_join.rs:1031]: Yes

File: datafusion/src/execution/context.rs
Comment (Dandandan) [reply datafusion/src/execution/context.rs:1308]: Also necessary to fix the tests - as the number of batches now also depend on the provided concurrency.

**Comment by Dandandan** [comment]
> This looks really cool @Dandandan. I suggest you add a test, if possible, that shows the plan with the repartition exec operation in order to prevent someone accidentally turning off this optimization during a refactor

Good idea! Added a test for this

File: datafusion/src/physical_plan/hash_aggregate.rs
Comment (alamb) [inline datafusion/src/physical_plan/hash_aggregate.rs:213]: 👍

**Review by alamb — APPROVED** [review]
I think it is looking great. Thanks @Dandandan

**Review by jorgecarleitao — APPROVED** [review]
I went through this and it looks great. Thanks a lot @Dandandan , great work. 💯 

:shipit:  from my side.

We may want to give some time in case someone else would like to go through this before merging.

❤️

**Comment by Dandandan** [comment]
@andygrove maybe? :)

**Comment by andygrove** [comment]
> @andygrove maybe? :)

I will make time this weekend to review this and take it for a spin!

**Comment by Dandandan** [comment]
Awesome, thanks @andygrove !

I have also some nice followup this weekend for more performance improvements for hash aggregates :D

**Review by andygrove — APPROVED** [review]
LGTM. I tested this out locally and confirmed that performance is much better for unpartitioned data, and about the same for partitioned data.

**Comment by andygrove** [comment]
@Dandandan Looks like there is a conflict that needs fixing

**Comment by andygrove** [comment]
I filed https://github.com/apache/arrow-datafusion/issues/346 for implementing this optimization in Ballista

**Comment by Dandandan** [comment]
Somehow coverage test seems to fail... But doesn't show what is failing

**Comment by alamb** [comment]
The https://github.com/apache/arrow-datafusion/pull/320/checks?check_run_id=2591763548 shows this buried in the logs (not at the end, annoyingly):

```
failures:

---- physical_plan::planner::tests::hash_agg_group_by_partitioned stdout ----
thread 'physical_plan::planner::tests::hash_agg_group_by_partitioned' panicked at 'assertion failed: formatted.contains(\"FinalPartitioned\")', datafusion/src/physical_plan/planner.rs:1051:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    physical_plan::planner::tests::hash_agg_group_by_partitioned
```

**Comment by Dandandan** [comment]
thanks @alamb will merge it now when it's green

**Comment by Dandandan** [comment]
Thanks all :tada:

**Pull request merged**
